server/src/main/java/org/elasticsearch/index/TimeSeriesIdGenerator.java

```diff
@@ -0,0 +1,334 @@
+ * <p>
+ * It's pretty useful to be able to "parse" the {@code _tsid} into the values
+ * of the dimensions that it encoded. It's useful to be able to do that without
+ * any index metadata.
+ * <h2>The {@code _tsid}</h2>
+ * Given these constraints the {@code _tsid} is <strong>just</strong> the field
+ * names and their values plus a little type information.
+ * <h2>Where does this happen?!</h2>
+ * We generate the {@code _tsid} on the node coordinating the index action so that
+ * it can be included in the routing.
+ */
+public final class TimeSeriesIdGenerator {
+    /**
+     * The maximum length of the tsid. The value itself comes from a range check in
+     * Lucene's writer for utf-8 doc values.
+     */
+    private static final int LIMIT = ByteBlockPool.BYTE_BLOCK_SIZE - 2;
+    /**
+     * Maximum length of the name of dimension. We picked this so that we could
+     * comfortable fit 16 dimensions inside {@link #LIMIT}.
+     */
+    private static final int DIMENSION_NAME_LIMIT = 512;
+    /**
+     * The maximum length of any single dimension. We picked this so that we could
+     * comfortable fit 16 dimensions inside {@link #LIMIT}. This should be quite
+     * comfortable given that dimensions are typically going to be less than a
+     * hundred bytes each, but we're being paranoid here.
+     */
+    private static final int DIMENSION_VALUE_LIMIT = 1024;
+
+    private final ObjectComponent root;
+
+    public TimeSeriesIdGenerator(ObjectComponent root) {
+        if (root == null) {
+            /*
+             * This can happen if an index is configured in time series mode
+             * without any mapping. It's fine - we'll add dimensions later.
+             * For now it'll make a generator that will fail to index any
+             * documents. Which is totally ok.
+             */
+            root = new ObjectComponent(Map.of());
+        }
+        root.collectDimensionNames("", name -> {
+            int bytes = UnicodeUtil.calcUTF16toUTF8Length(name, 0, name.length());
+            if (bytes > DIMENSION_NAME_LIMIT) {
+                throw new IllegalArgumentException(
+                    "Dimension name must be less than [" + DIMENSION_NAME_LIMIT + "] bytes but [" + name + "] was [" + bytes + "]"
+                );
+            }
+        });
+        this.root = root;
+    }
+
+    @Override
+    public String toString() {
+        return "extract dimensions using " + root;
+    }
+
+    /**
+     * Build the tsid from the {@code _source}. See class docs for more on what it looks like and why.
+     */
+    public BytesReference generate(XContentParser parser) throws IOException {
+        List<Map.Entry<String, CheckedConsumer<StreamOutput, IOException>>> values = new ArrayList<>();
+        parser.nextToken();
+        root.extract(values, "", parser);
+        if (values.isEmpty()) {
+            List<String> dimensionNames = new ArrayList<>();
+            root.collectDimensionNames("", dimensionNames::add);
```

Comment:
You can probably move this to constructor, you have to run collectDimensionNames anyway.

Reply:
I didn't want keep the string sitting around though. In case there are zillions of these things.